File: typescript/day3/main.ts
import { file } from "../util/main";

const report = file("./data.txt");
// const report = `00100
// 11110
// 10110
// 10111
// 10101
// 01111
// 00111
// 11100
// 10000
// 11001
// 00010
// 01010`;


function generateNewArray(arr: number[][]): number[][] {
  let newArr = [];
  let temp = [];
  let index = 0;
  while(index < arr[0].length){
    for(let i = 0; i < arr.length; i++){
      temp.push(arr[i][index])
    }
    newArr.push(temp);
    temp = [];
    index += 1;
  }

  return newArr;
}


function bitOperation(arr: number[][]) {
  const data = {gamma: [], eplison: []};

  const temp = arr.map((v, _) => {
    const count = {0: 0, 1: 0};
    v.forEach(value => {
      if(value === 0){
	count[0]++
      } else if(value === 1) {
	count[1]++;
      }
    })
    return count;
  }) 

  temp.forEach(v => {
    if(v[0] > v[1]){
      data.gamma.push(0);
      data.eplison.push(1);
    } else {
      data.gamma.push(1);
      data.eplison.push(0);
    }
  })

  return data;
}

const lines = report.split("\n")
  .map((v) => v.split("").map(Number));

const data = bitOperation(generateNewArray(lines))
const gamma = parseInt(data.gamma.join(""), 2);
const eplison = parseInt(data.eplison.join(""), 2);

console.log(`The power consumption is ${gamma*eplison}`);

File: golang/day3/main.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

func report() string {
	data, err := ioutil.ReadFile("./data.txt")
	if err != nil {
		panic(err)
	}

	return string(data)
}

func generateNewArray(arr [][]int) [][]int {
	newArr := make([][]int, len(arr[0]))

	for i := 0; i < len(arr[0]); i++ {
		var temp []int
		for j := 0; j < len(arr); j++ {
			temp = append(temp, arr[j][i])
		}
		newArr[i] = temp
	}
	return newArr
}

func bitOperation(arr [][]int) (gamma, epsilon int) {
	for _, v := range arr {
		count := map[int]int{0: 0, 1: 0}
		for _, val := range v {
			count[val]++
		}
		if count[0] > count[1] {
			gamma = gamma<<1 | 0
			epsilon = epsilon<<1 | 1
		} else {
			gamma = gamma<<1 | 1
			epsilon = epsilon<<1 | 0
		}
	}

	return gamma, epsilon
}

func main() {
	data := strings.Split(report(), "\n")
	var arr [][]int

	for _, row := range data {
		if row == "" {
			continue
		}
		var newRow []int
		for _, val := range strings.Split(row, "") {
			num, _ := strconv.Atoi(val)
			newRow = append(newRow, num)
		}
		arr = append(arr, newRow)
	}

	if len(arr) == 0 {
		fmt.Println("No data found.")
		return
	}

	newArr := generateNewArray(arr)
	gamma, epsilon := bitOperation(newArr)
	powerConsumption := gamma * epsilon

	fmt.Printf("The power consumption is %d\n", powerConsumption)
}
